nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/bulk/BulkIndexComputation.java

```diff
@@ -0,0 +1,199 @@
+        codec = BulkCodecs.getDataBucketCodec();
+        context.setTimer("flush", System.currentTimeMillis() + flushIntervalMs);
+    }
+
+    @Override
+    public void processTimer(ComputationContext context, String key, long timestamp) {
+        if (abort) {
+            context.askForTermination();
+            log.error("Terminate computation due to previous error");
+            return;
+        }
+        if (updates) {
+            // flush is sync because bulkProcessor is initialized with setConcurrentRequests(0)
+            bulkProcessor.flush();
+            context.askForCheckpoint();
+            updates = false;
+        }
+        context.setTimer("flush", System.currentTimeMillis() + flushIntervalMs);
+    }
+
+    @Override
+    public void processRecord(ComputationContext context, String inputStream, Record record) {
+        DataBucket in = codec.decode(record.getData());
+        BulkRequest bulkRequest = decodeRequest(in);
+        for (DocWriteRequest request : bulkRequest.requests()) {
+            bulkProcessor.add(request);
+        }
+        AbstractBulkComputation.updateStatusProcessed(context, in.getCommandId(), in.getCount());
+        updates = true;
+    }
+
+    @Override
+    public void destroy() {
+        if (bulkProcessor != null) {
+            bulkProcessor.close();
+            bulkProcessor = null;
+        }
+    }
+
+    protected ESClient getESClient() {
+        ElasticSearchAdmin esa = Framework.getService(ElasticSearchAdmin.class);
+        return esa.getClient();
+    }
+
+    protected BulkRequest decodeRequest(DataBucket bucket) {
+        BulkRequest ret = new BulkRequest();
+        ByteBuffer buffer = ByteBuffer.wrap(bucket.getData());
+        StreamInput in = new ByteBufferStreamInput(buffer);
+        try {
+            ret.readFrom(in);
+            return ret;
+        } catch (IOException e) {
+            throw new NuxeoException("Cannot load elastic bulk request from: " + bucket);
+        }
+    }
+
+    // -------------------------------------------------------------------------------------
+    // Elasticsearch bulk processor listener
+    // the following methods are called from a different thread than the computation
+    @Override
+    public void beforeBulk(long executionId, BulkRequest request) {
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Creating elasticsearch bulk %s with %d action", executionId,
+                    request.numberOfActions()));
+        }
+    }
+
+    @Override
+    public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("After bulk: %s, actions: %d, status: %s", executionId, request.numberOfActions(),
+                    response.status()));
+        }
+        if (!response.hasFailures()) {
+            return;
+        }
+        AtomicBoolean inError = new AtomicBoolean(false);
```

Comment:
Unless you really want thread-safety, a simple Apache Lang3 `MutableBoolean` is cheaper.
Although there's probably a better Stream-like way of doing this...